File: DemoObjc/DemoObjc/SwitchChainViewController.m
//
//  SwitchChainViewController.m
//  DemoObjc
//
//  Created by link on 2022/6/21.
//

#import "SwitchChainViewController.h"

@import ParticleNetworkBase;

@interface SwitchChainViewController()
@property (nonatomic, strong) UITableView *tableView;
@property (nonatomic, strong) NSMutableArray *data;
@end

@implementation SwitchChainViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    
    [self configureData];
    [self configureTableView];
}

- (void)configureData {
    self.data = [[NSMutableArray alloc] init];
    NSMutableDictionary *solana = [[NSMutableDictionary alloc] init];
    NSMutableDictionary *ethereum = [[NSMutableDictionary alloc] init];
    NSMutableDictionary *bsc = [[NSMutableDictionary alloc] init];
    NSMutableDictionary *polygon = [[NSMutableDictionary alloc] init];
    NSMutableDictionary *avalanche = [[NSMutableDictionary alloc] init];
    NSMutableDictionary *fantom = [[NSMutableDictionary alloc] init];
    NSMutableDictionary *arbitrum = [[NSMutableDictionary alloc] init];
    NSMutableDictionary *moonBeam = [[NSMutableDictionary alloc] init];
    NSMutableDictionary *moonRiver = [[NSMutableDictionary alloc] init];
    NSMutableDictionary *heco = [[NSMutableDictionary alloc] init];
    NSMutableDictionary *aurora = [[NSMutableDictionary alloc] init];
    NSMutableDictionary *harmony = [[NSMutableDictionary alloc] init];
    NSMutableDictionary *kcc = [[NSMutableDictionary alloc] init];
    NSMutableDictionary *platon = [[NSMutableDictionary alloc] init];
    NSMutableDictionary *optimism = [[NSMutableDictionary alloc] init];
    NSMutableDictionary *tron = [[NSMutableDictionary alloc] init];
    NSMutableDictionary *okc = [[NSMutableDictionary alloc] init];
    NSMutableDictionary *thunderCore = [[NSMutableDictionary alloc] init];
    NSMutableDictionary *cronos = [[NSMutableDictionary alloc] init];
    NSMutableDictionary *oasisEmerald = [[NSMutableDictionary alloc] init];
    NSMutableDictionary *gnosis = [[NSMutableDictionary alloc] init];
    NSMutableDictionary *celo = [[NSMutableDictionary alloc] init];
    NSMutableDictionary *klaytn = [[NSMutableDictionary alloc] init];
    NSMutableDictionary *scroll = [[NSMutableDictionary alloc] init];
    NSMutableDictionary *zkSync = [[NSMutableDictionary alloc] init];
    NSMutableDictionary *metis = [[NSMutableDictionary alloc] init];
    
    [solana setObject:@[@"mainnet", @"testnet", @"devnet"] forKey:@"solana"];
    [ethereum setObject:@[@"mainnet", @"goerli"] forKey:@"ethereum"];
    [bsc setObject:@[@"mainnet", @"testnet"] forKey:@"bsc"];
    [polygon setObject:@[@"mainnet", @"mumbai"] forKey:@"polygon"];
    
    [avalanche setObject:@[@"mainnet", @"testnet"] forKey:@"avalanche"];
    [fantom setObject:@[@"mainnet", @"testnet"] forKey:@"fantom"];
    [arbitrum setObject:@[@"one", @"nova", @"goerli"] forKey:@"arbitrum"];
    [moonBeam setObject:@[@"mainnet", @"testnet"] forKey:@"moonbeam"];
    
    [moonRiver setObject:@[@"mainnet", @"testnet"] forKey:@"moonriver"];
    [heco setObject:@[@"mainnet", @"testnet"] forKey:@"heco"];
    [aurora setObject:@[@"mainnet", @"testnet"] forKey:@"aurora"];
    [harmony setObject:@[@"mainnet", @"testnet"] forKey:@"harmony"];
    
    [kcc setObject:@[@"mainnet", @"testnet"] forKey:@"kcc"];
    [platon setObject:@[@"mainnet", @"testnet"] forKey:@"platon"];
    [optimism setObject:@[@"mainnet", @"goerli"] forKey:@"optimism"];
    [tron setObject:@[@"mainnet", @"shasta", @"nile"] forKey:@"tron"];
    
    [okc setObject:@[@"mainnet", @"testnet"] forKey:@"okc"];
    [thunderCore setObject:@[@"mainnet", @"testnet"] forKey:@"thunderCore"];
    [cronos setObject:@[@"mainnet", @"testnet"] forKey:@"cronos"];
    [oasisEmerald setObject:@[@"mainnet", @"testnet"] forKey:@"oasisEmerald"];
    [gnosis setObject:@[@"mainnet", @"testnet"] forKey:@"gnosis"];
    [celo setObject:@[@"mainnet", @"testnet"] forKey:@"celo"];
    [klaytn setObject:@[@"mainnet", @"testnet"] forKey:@"klaytn"];
    [scroll setObject:@[@"testnetL1", @"testnetL2"] forKey:@"scroll"];
    [zkSync setObject:@[@"mainnet", @"testnet"] forKey:@"zkSync"];
    [metis setObject:@[@"mainnet", @"goerli"] forKey:@"metis"];
    
    [self.data addObjectsFromArray:@[solana, ethereum, bsc, polygon, avalanche, fantom, arbitrum, moonBeam, moonRiver, heco, aurora, harmony, kcc, platon, optimism, tron]];
}

- (void)configureTableView {
    self.tableView = [[UITableView alloc] initWithFrame:CGRectZero style:UITableViewStyleGrouped];
    [self.view addSubview:self.tableView];
    
    self.tableView.translatesAutoresizingMaskIntoConstraints = NO;
    [self.tableView.leftAnchor constraintEqualToAnchor:self.view.leftAnchor].active = YES;
    [self.tableView.rightAnchor constraintEqualToAnchor:self.view.rightAnchor].active = YES;
    [self.tableView.topAnchor constraintEqualToAnchor:self.view.topAnchor].active = YES;
    [self.tableView.bottomAnchor constraintEqualToAnchor:self.view.bottomAnchor].active = YES;
    
    
    [self.tableView registerClass:[UITableViewCell class] forCellReuseIdentifier:NSStringFromClass([UITableViewCell class])];
    
    self.tableView.delegate = self;
    self.tableView.dataSource = self;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    NSMutableDictionary *dict = self.data[section];
    NSArray *array = dict.allValues.firstObject;
    return array.count;
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return self.data.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:NSStringFromClass([UITableViewCell class]) forIndexPath:indexPath];
    cell.selectionStyle = UITableViewCellSelectionStyleNone;
    NSMutableDictionary *dict = self.data[indexPath.section];
    NSArray *array = [[dict allValues] firstObject];
    NSString *network = array[indexPath.row];
    cell.textLabel.text = network;
    return cell;
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
    NSMutableDictionary *dict = self.data[section];
    return [[dict allKeys] firstObject];
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    NSMutableDictionary *dict = self.data[indexPath.section];
    NSArray *array = [[dict allValues] firstObject];
    NSString *network = [array[indexPath.row] lowercaseString];
    NSString *name = [[[dict allKeys] firstObject] lowercaseString];
    
    NSLog(@"%@, %@", name, network);
    ChainInfo *chainInfo;
    if ([name  isEqual: @"solana"]) {
        if ([network  isEqual: @"mainnet"]) {
            chainInfo = [ChainInfo solana:SolanaNetworkMainnet];
        } else if ([network  isEqual: @"testnet"]) {
            chainInfo = [ChainInfo solana:SolanaNetworkTestnet];
        } else if ([network  isEqual: @"devnet"]) {
            chainInfo = [ChainInfo solana:SolanaNetworkDevnet];
        }
    } else if ([name  isEqual: @"ethereum"]) {
        if ([network  isEqual: @"mainnet"]) {
            chainInfo = [ChainInfo ethereum:EthereumNetworkMainnet];
        } else if ([network  isEqual: @"goerli"]) {
            chainInfo = [ChainInfo ethereum:EthereumNetworkGoerli];
        }
    } else if ([name  isEqual: @"bsc"]) {
        if ([network  isEqual: @"mainnet"]) {
            chainInfo = [ChainInfo bsc:BscNetworkMainnet];
        } else if ([network  isEqual: @"testnet"]) {
            chainInfo = [ChainInfo bsc:BscNetworkTestnet];
        }
    } else if ([name  isEqual: @"polygon"]) {
        if ([network  isEqual: @"mainnet"]) {
            chainInfo = [ChainInfo polygon:PolygonNetworkMainnet];
        } else if ([network  isEqual: @"mumbai"]) {
            chainInfo = [ChainInfo polygon:PolygonNetworkMumbai];
        }
    } else if ([name  isEqual: @"avalanche"]) {
        if ([network  isEqual: @"mainnet"]) {
            chainInfo = [ChainInfo avalanche:AvalancheNetworkMainnet];
        } else if ([network  isEqual: @"testnet"]) {
            chainInfo = [ChainInfo avalanche:AvalancheNetworkTestnet];
        }
    } else if ([name  isEqual: @"fantom"]) {
        if ([network  isEqual: @"mainnet"]) {
            chainInfo = [ChainInfo fantom:FantomNetworkMainnet];
        } else if ([network  isEqual: @"testnet"]) {
            chainInfo = [ChainInfo fantom:FantomNetworkTestnet];
        }
    } else if ([name  isEqual: @"arbitrum"]) {
        if ([network  isEqual: @"one"]) {
            chainInfo = [ChainInfo arbitrum:ArbitrumNetworkOne];
        } else if ([network  isEqual: @"nova"]) {
            chainInfo = [ChainInfo arbitrum:ArbitrumNetworkNova];
        } else if ([network  isEqual: @"goerli"]) {
            chainInfo = [ChainInfo arbitrum:ArbitrumNetworkGoerli];
        }
    } else if ([name  isEqual: @"moonbeam"]) {
        if ([network  isEqual: @"mainnet"]) {
            chainInfo = [ChainInfo moonbeam:MoonbeamNetworkMainnet];
        } else if ([network  isEqual: @"testnet"]) {
            chainInfo = [ChainInfo moonbeam:MoonbeamNetworkTestnet];
        }
    } else if ([name  isEqual: @"moonriver"]) {
        if ([network  isEqual: @"mainnet"]) {
            chainInfo = [ChainInfo moonriver:MoonriverNetworkMainnet];
        } else if ([network  isEqual: @"testnet"]) {
            chainInfo = [ChainInfo moonriver:MoonriverNetworkTestnet];
        }
    } else if ([name  isEqual: @"heco"]) {
        if ([network  isEqual: @"mainnet"]) {
            chainInfo = [ChainInfo heco:HecoNetworkMainnet];
        }
    } else if ([name  isEqual: @"aurora"]) {
        if ([network  isEqual: @"mainnet"]) {
            chainInfo = [ChainInfo aurora:AuroraNetworkMainnet];
        } else if ([network  isEqual: @"testnet"]) {
            chainInfo = [ChainInfo aurora:AuroraNetworkTestnet];
        }
    } else if ([name  isEqual: @"harmony"]) {
        if ([network  isEqual: @"mainnet"]) {
            chainInfo = [ChainInfo harmony:HarmonyNetworkMainnet];
        } else if ([network  isEqual: @"testnet"]) {
            chainInfo = [ChainInfo harmony:HarmonyNetworkTestnet];
        }
    } else if ([name  isEqual: @"platon"]) {
        if ([network  isEqual: @"mainnet"]) {
            chainInfo = [ChainInfo platON:PlatONNetworkMainnet];
        } else if ([network  isEqual: @"testnet"]) {
            chainInfo = [ChainInfo platON:PlatONNetworkTestnet];
        }
    } else if ([name  isEqual: @"kcc"]) {
        if ([network  isEqual: @"mainnet"]) {
            chainInfo = [ChainInfo kcc:KccNetworkMainnet];
        } else if ([network  isEqual: @"testnet"]) {
            chainInfo = [ChainInfo kcc:KccNetworkTestnet];
        }
    } else if ([name  isEqual: @"optimism"]) {
        if ([network  isEqual: @"mainnet"]) {
            chainInfo = [ChainInfo optimism:OptimismNetworkMainnet];
        } else if ([network  isEqual: @"goerli"]) {
            chainInfo = [ChainInfo optimism:OptimismNetworkGoerli];
        }
    } else if ([name  isEqual:@"tron"]) {
        if ([network  isEqual: @"mainnet"]) {
            chainInfo = [ChainInfo tron:TronNetworkMainnet];
        } else if ([network  isEqual: @"shasta"]) {
            chainInfo = [ChainInfo tron:TronNetworkShasta];
        } else if ([network  isEqual: @"nile"]) {
            chainInfo = [ChainInfo tron:TronNetworkNile];
        }
    } else if ([name  isEqual: @"okc"]) {
        if ([network  isEqual: @"mainnet"]) {
            chainInfo = [ChainInfo okc:OKCNetworkMainnet];
        } else if ([network  isEqual: @"testnet"]) {
            chainInfo = [ChainInfo okc:OKCNetworkTestnet];
        }
    } else if ([name  isEqual: @"thunderCore"]) {
        if ([network  isEqual: @"mainnet"]) {
            chainInfo = [ChainInfo thunderCore:ThunderCoreNetworkMainnet];
        } else if ([network  isEqual: @"testnet"]) {
            chainInfo = [ChainInfo thunderCore:ThunderCoreNetworkTestnet];
        }
    } else if ([name  isEqual: @"cronos"]) {
        if ([network  isEqual: @"mainnet"]) {
            chainInfo = [ChainInfo cronos:CronosNetworkMainnet];
        } else if ([network  isEqual: @"testnet"]) {
            chainInfo = [ChainInfo cronos:CronosNetworkTestnet];
        }
    } else if ([name  isEqual: @"oasisEmerald"]) {
        if ([network  isEqual: @"mainnet"]) {
            chainInfo = [ChainInfo oasisEmerald:OasisEmeraldNetworkMainnet];
        } else if ([network  isEqual: @"testnet"]) {
            chainInfo = [ChainInfo oasisEmerald:OasisEmeraldNetworkTestnet];
        }
    } else if ([name  isEqual: @"gnosis"]) {
        if ([network  isEqual: @"mainnet"]) {
            chainInfo = [ChainInfo gnosis:GnosisNetworkMainnet];
        } else if ([network  isEqual: @"testnet"]) {
            chainInfo = [ChainInfo gnosis:GnosisNetworkTestnet];
        }
    } else if ([name  isEqual: @"celo"]) {
        if ([network  isEqual: @"mainnet"]) {
            chainInfo = [ChainInfo celo:CeloNetworkMainnet];
        } else if ([network  isEqual: @"testnet"]) {
            chainInfo = [ChainInfo celo:CeloNetworkTestnet];
        }
    } else if ([name  isEqual: @"klaytn"]) {
        if ([network  isEqual: @"mainnet"]) {
            chainInfo = [ChainInfo klaytn:KlaytnNetworkMainnet];
        } else if ([network  isEqual: @"testnet"]) {
            chainInfo = [ChainInfo klaytn:KlaytnNetworkTestnet];
        }
    } else if ([name  isEqual: @"scroll"]) {
        if ([network  isEqual: @"testnet"]) {
            chainInfo = [ChainInfo scroll:ScrollNetworkTestnet];
        }
    } else if ([name  isEqual: @"zkSync"]) {
        if ([network  isEqual: @"mainnet"]) {
            chainInfo = [ChainInfo zkSync:ZkSyncNetworkMainnet];
        } else if ([network  isEqual: @"testnet"]) {
            chainInfo = [ChainInfo zkSync:ZkSyncNetworkTestnet];
        }
    } else if ([name  isEqual: @"metis"]) {
        if ([network  isEqual: @"mainnet"]) {
            chainInfo = [ChainInfo metis:MetisNetworkMainnet];
        } else if ([network  isEqual: @"goerli"]) {
            chainInfo = [ChainInfo metis:MetisNetworkGoerli];
        }
    }
    
    [ParticleNetwork setChainInfo:chainInfo];
    
    NSString *message = [NSString stringWithFormat:@"current network is %@ - %@", name, network];
    
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Switch network" message:message preferredStyle:UIAlertControllerStyleAlert];
    [alert addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        self.selectHandler();
        [self dismissViewControllerAnimated:true completion:false];
    }]];
    
    [self presentViewController:alert animated:YES completion:nil];
    
}


- (void)dealloc {
    NSLog(@"%@ dealloc", NSStringFromClass([SwitchChainViewController class]));
}
@end

File: Demo/Demo/SwitchChainViewController.swift
//
//  SwitchChainViewController.swift
//  Demo
//
//  Created by link on 2022/6/6.
//  Copyright © 2022 ParticleNetwork. All rights reserved.
//

import Foundation
import ParticleAuthService
import ParticleNetworkBase
import RxSwift
import UIKit

typealias Chain = ParticleNetwork.ChainName
typealias SolanaNetwork = ParticleNetwork.SolanaNetwork
typealias EthereumNetwork = ParticleNetwork.EthereumNetwork
typealias BscNetwork = ParticleNetwork.BscNetwork
typealias PolygonNetwork = ParticleNetwork.PolygonNetwork
typealias AvalancheNetwork = ParticleNetwork.AvalancheNetwork
typealias FantomNetwork = ParticleNetwork.FantomNetwork
typealias ArbitrumNetwork = ParticleNetwork.ArbitrumNetwork
typealias MoonbeamNetwork = ParticleNetwork.MoonbeamNetwork
typealias MoonriverNetwork = ParticleNetwork.MoonriverNetwork
typealias HecoNetwork = ParticleNetwork.HecoNetwork
typealias AuroraNetwork = ParticleNetwork.AuroraNetwork
typealias HarmonyNetwork = ParticleNetwork.HarmonyNetwork
typealias KccNetwork = ParticleNetwork.KccNetwork

class SwitchChainViewController: UIViewController {
    let bag = DisposeBag()
    var selectHandler: (() -> Void)?
    let tableView = UITableView(frame: .zero, style: .grouped)

    var data: [[String: [String]]] = []

    override func viewDidLoad() {
        super.viewDidLoad()
        configureData()
        configureTableView()
    }

    func configureData() {
        data.append([Chain.solana(.mainnet).nameString: [
            SolanaNetwork.mainnet.rawValue, SolanaNetwork.testnet.rawValue, SolanaNetwork.devnet.rawValue
        ]])
        data.append([Chain.ethereum(.mainnet).nameString: [
            EthereumNetwork.mainnet.rawValue, EthereumNetwork.kovan.rawValue
        ]])
        data.append([Chain.bsc(.mainnet).nameString: [
            BscNetwork.mainnet.rawValue, BscNetwork.testnet.rawValue
        ]])
        data.append([Chain.polygon(.mainnet).nameString: [
            PolygonNetwork.mainnet.rawValue, PolygonNetwork.mumbai.rawValue
        ]])
        data.append([Chain.avalanche(.mainnet).nameString: [
            AvalancheNetwork.mainnet.rawValue, AvalancheNetwork.testnet.rawValue
        ]])
        data.append([Chain.fantom(.mainnet).nameString: [
            FantomNetwork.mainnet.rawValue, FantomNetwork.testnet.rawValue
        ]])
        data.append([Chain.arbitrum(.mainnet).nameString: [
            ArbitrumNetwork.mainnet.rawValue, ArbitrumNetwork.testnet.rawValue
        ]])
        data.append([Chain.moonbeam(.mainnet).nameString: [
            MoonbeamNetwork.mainnet.rawValue, MoonbeamNetwork.testnet.rawValue
        ]])
        data.append([Chain.moonriver(.mainnet).nameString: [
            MoonriverNetwork.mainnet.rawValue, MoonriverNetwork.testnet.rawValue
        ]])
        data.append([Chain.heco(.mainnet).nameString: [
            HecoNetwork.mainnet.rawValue, HecoNetwork.testnet.rawValue
        ]])
        data.append([Chain.aurora(.mainnet).nameString: [
            AuroraNetwork.mainnet.rawValue, AuroraNetwork.testnet.rawValue
        ]])
        data.append([Chain.harmony(.mainnet).nameString: [
            HarmonyNetwork.mainnet.rawValue, HarmonyNetwork.testnet.rawValue
        ]])
        data.append([Chain.kcc(.mainnet).nameString: [
            KccNetwork.mainnet.rawValue, KccNetwork.testnet.rawValue
        ]])
    }

    func configureTableView() {
        view.addSubview(tableView)

        tableView.snp.makeConstraints { make in
            make.top.bottom.left.right.equalToSuperview()
        }

        tableView.register(UITableViewCell.self, forCellReuseIdentifier: NSStringFromClass(UITableViewCell.self))

        tableView.delegate = self
        tableView.dataSource = self
    }
}

extension SwitchChainViewController: UITableViewDataSource {
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return data[section].values.first?.count ?? 0
    }

    func numberOfSections(in tableView: UITableView) -> Int {
        return data.count
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: NSStringFromClass(UITableViewCell.self), for: indexPath)
        let network = data[indexPath.section].values.first?[indexPath.row] ?? ""
        cell.textLabel?.text = network
        cell.selectionStyle = .none
        return cell
    }

    func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        data[section].keys.first
    }
}

extension SwitchChainViewController: UITableViewDelegate {
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let network = data[indexPath.section].values.first?[indexPath.row] ?? ""
        let name = data[indexPath.section].keys.first ?? ""

        var chainName: Chain
        switch name {
        case Chain.solana(.mainnet).nameString:
            chainName = .solana(SolanaNetwork(rawValue: network)!)
        case Chain.ethereum(.mainnet).nameString:
            chainName = .ethereum(EthereumNetwork(rawValue: network)!)
        case Chain.bsc(.mainnet).nameString:
            chainName = .bsc(BscNetwork(rawValue: network)!)
        case Chain.polygon(.mainnet).nameString:
            chainName = .polygon(PolygonNetwork(rawValue: network)!)
        case Chain.avalanche(.mainnet).nameString:
            chainName = .avalanche(AvalancheNetwork(rawValue: network)!)
        case Chain.fantom(.mainnet).nameString:
            chainName = .fantom(FantomNetwork(rawValue: network)!)
        case Chain.arbitrum(.mainnet).nameString:
            chainName = .arbitrum(ArbitrumNetwork(rawValue: network)!)
        case Chain.moonbeam(.mainnet).nameString:
            chainName = .moonbeam(MoonbeamNetwork(rawValue: network)!)
        case Chain.moonriver(.mainnet).nameString:
            chainName = .moonriver(MoonriverNetwork(rawValue: network)!)
        case Chain.heco(.mainnet).nameString:
            chainName = .heco(HecoNetwork(rawValue: network)!)
        case Chain.aurora(.mainnet).nameString:
            chainName = .aurora(AuroraNetwork(rawValue: network)!)
        case Chain.harmony(.mainnet).nameString:
            chainName = .harmony(HarmonyNetwork(rawValue: network)!)
        case Chain.kcc(.mainnet).nameString:
            chainName = .kcc(KccNetwork(rawValue: network)!)
        default:
            chainName = .ethereum(.mainnet)
        }
        if ParticleAuthService.isLogin() {
            ParticleAuthService.setChainName(chainName).subscribe {
                [weak self] result in
                    guard let self = self else { return }
                    switch result {
                    case .failure(let error):
                        print(error)
                    case .success(let userInfo):
                        print(String(describing: userInfo))

                        let alert = UIAlertController(title: "Switch network", message: "current network is \(name) - \(network)", preferredStyle: .alert)
                        alert.addAction(UIAlertAction(title: "OK", style: .default, handler: { _ in

                            if let selectHandler = self.selectHandler {
                                selectHandler()
                            }
                            self.dismiss(animated: true)
                        }))

                        self.present(alert, animated: true)
                    }
            }.disposed(by: bag)
        } else {
            ParticleNetwork.setChainName(chainName)

            let alert = UIAlertController(title: "Switch network", message: "current network is \(name) - \(network)", preferredStyle: .alert)
            alert.addAction(UIAlertAction(title: "OK", style: .default, handler: { _ in

                if let selectHandler = self.selectHandler {
                    selectHandler()
                }
                self.dismiss(animated: true)
            }))

            present(alert, animated: true)
        }
    }
}
